marlin-succinct/protocol/src/batch_verifier.rs

```diff
@@ -0,0 +1,101 @@
+/********************************************************************************************
+
+This source file implements batch zk-proof primitives.
+
+This primitive is the top-level aggregate of Marlin zk-proof format that provides means of
+optimized batch verification of the individual constituent zk-proofs.
+
+*********************************************************************************************/
+
+use rand_core::RngCore;
+use algebra::PairingEngine;
+use circuits::index::Index;
+use oracle::rndoracle::ProofError;
+pub use super::prover::{ProverProof, RandomOracles};
+pub use super::batch_prover::BatchProof;
+
+impl<E: PairingEngine> BatchProof<E>
+{
+    // This function verifies the batch of zk-proofs
+    //     index: Index
+    //     rng: randomness source context
+    //     RETURN: verification status
+    pub fn verify
+    (
+        &self,
+        index: &Index<E>,
+        rng: &mut dyn RngCore
+    ) -> Result<bool, ProofError>
+    {
+        let mut batch = vec![Vec::new(), Vec::new(), Vec::new()];
+        for proof in self.batch.iter()
+        {
+            match proof
```

Comment:
I think you can use the `?` operator to make this code less nested: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html